archive/miu_language/decision_suf.lean

```diff
@@ -0,0 +1,382 @@
+      apply derivable.r1,
+      rw repeat_pow_minus_append,
+      exact (der_cons_repeat m), },
+  have hw₃ : derivable (M::(repeat I c) ++ repeat U ((2^m-c)/3) ++ repeat U ((2^m-c)/3 % 2)),
+  { apply der_cons_repeat_I_repeat_U_append_of_der_cons_repeat_I_append c ((2^m-c)/3),
+    { exact h, }, /- `c` is 1 or 2 (mod 3) -/
+      convert hw₂, -- now we must show `c + 3 * ((2 ^ m - c) / 3) = 2 ^ m`
+      rw nat.mul_div_cancel',
+      { exact add_sub_of_le hm.1, },
+      exact (modeq.modeq_iff_dvd' hm.1).mp hm.2.symm, },
+  rw [append_assoc, ←repeat_add _ _] at hw₃,
+  cases add_mod2 ((2^m-c)/3) with t ht,
+  rw ht at hw₃,
+  exact der_of_der_append_repeat_U_even hw₃,
+end
+
+
+
+/-!
+### `decstr` is a sufficient condition
+
+The remainder of this file sets up the proof that `dectstr en` is sufficent to ensure
+`derivable en`. Decidability of `derivable en` is an easy consequence.
+
+The proof proceeds by induction on the `count U` of `en`.
+
+We tackle first the base case of the induction. This requires auxiliary results giving
+conditions under which  `count I ys = length ys`.
+-/
+
+
+/--
+If an `miustr` has a zero `count U` and contains no `M`, then its `count I` is its length.
+-/
+lemma count_I_eq_length_of_count_U_zero_and_neg_mem {ys : miustr} (hu : count U ys = 0)
+  (hm : M ∉ ys) : count I ys = length ys :=
+begin
+  induction ys with x xs hxs,
+  { refl, },
+  cases x,
+  { exfalso, exact hm (mem_cons_self M xs), }, -- case `x = M` gives a contradiction.
+  { rw [count_cons, if_pos (rfl),length,succ_eq_add_one,succ_inj'], -- case `x = I`
+    apply hxs,
+    { simpa only [count], },
+    simp only [mem_cons_iff,false_or] at hm, exact hm, },
+  exfalso, simp only [count,countp_cons_of_pos] at hu,  -- case `x = U` gives a contradiction.
+  exact succ_ne_zero _ hu,
+end
+
+/--
+`base_case_suf` is the base case of the sufficiency result.
+-/
+lemma base_case_suf (en : miustr) (h : decstr en) (hu : count U en = 0) : derivable en :=
+begin
+  rcases h with ⟨⟨mhead, nmtail⟩, hi ⟩,
+  have : en ≠ nil,
+  { intro k, simp [k,count] at hi, contradiction, },
+  rcases (exists_cons_of_ne_nil this) with ⟨y,ys,rfl⟩,
+  rw head at mhead,
+  rw mhead at *,
+  suffices  : ∃ c, repeat I c = ys ∧ (c % 3 = 1 ∨ c % 3 = 2),
+  { rcases this with ⟨c, hysr, hc⟩,
+    rw ←hysr,
+    exact der_repeat_I_of_mod3 c hc, },
+  simp only [count] at *,
+  use (count I ys),
+  split,
+  { apply repeat_count_eq_of_count_eq_length,
+    apply count_I_eq_length_of_count_U_zero_and_neg_mem,
+    { exact hu, },
+    exact nmtail, },
+  exact hi,
+end
+
+
+
+
+/-!
+Before continuing to the proof of the induction step, we need other auxiliary results that
+relate to `count U`.
+-/
+lemma mem_of_count_U_eq_succ {xs : miustr} {k : ℕ} (h : count U xs = succ k) : U ∈ xs :=
+begin
+  induction xs with z zs hzs,
+  { exfalso, rw count at h, contradiction, },
+  simp only [mem_cons_iff],
+  cases z,
+    repeat -- cases `z = M` and `z=I`
```

Comment:
Is this multiple subgoals? `{}`?

Reply:
I'm not certain I know how to format tactics that use `repeat`, `any_goals`, etc. I've made what I hope are acceptable revisions. Please check:

I have `any_goals` on on line 76 of `decision_nec.lean` and the above use of `repeat` now on line 312 of `decision_suf.lean`.

Reply:
I made an indentation suggestion here, otherwise it looks good. My rule of thumb: don't change the indentation for a tactic call except when entering a `{ ... }`. In that case, increase indentation by two.